Update FilePicker to 0.8.x

n3dst4 opened the pull request:
Resolves #356 (I think)

UFOMelkor reviewed (changes requested):
Thank you for your work!

While your changes look good to me, it is sincerely not enough to close #938 

We try to follow the structure of the source code as close as possible, especially the order of the elements makes reviewing a lot easier.

Might I ask you to address this as well?

There are some points where the current types diverge from foundy:

- `constructor` is missing
- `request` should be `string|undefined`
- `type` should be `string|undefined`
- `field` should be `HTMLElement|undefined`
- `button` should be `HTMLElement|undefined`
- `displayMode` is missing a defaultValue
- `extensions` looks like it can be `undefined`

n3dst4 commented:
Updated! Thank you for taking the time to review my PR.

`callback` is in the same order as source now.

I've addressed the other types you pointed out.

I've changed the default for`LAST_DISPLAY_MODE` to simply `'list'` - previous implementation had it as `FilePicker.DISPLAY_MODES.List` which isn't actually a meaningful value.

UFOMelkor commented on src/foundry/foundry.js/applications/filePicker.d.ts:
```diff
@@ -81,7 +92,7 @@ declare class FilePicker<P extends FilePicker.Options = FilePicker.Options> exte
+  static parseS3URL(key: string): { bucket: string | null; keyPrefix: string };
 
   /**
    * @override
```
For the methods below:

- `browse()`:
    Please document the default value of the `options` parameter
    Return type should be `Promise<FilePicker.Result | undefined>`
- `static browse()`:
    Please document the default value of the `options` parameter
- `configurePath()`:
    Please document the default value of the `options` parameter
- `createDirectory()`:
    Please document the default value of the `options` parameter
- `upload()`:
    Please document the default value of the `options` parameter
- `render()`:
    I guess the return type should be something like `void | ReturnType<this['browse']>`
- `_onRequestTarget()`:
    I guess the return type should be something like `void | ReturnType<this['browse']>`
- `_onLazyLoadImages()`:
    Parameters can be grabbed from `SidebarTab#_onLazyLoadImages`

n3dst4 replied:
@UFOMelkor Work is in progress. I have to say, most of these things are not 0.8.x-related changes but I'm happy to help get the quality level up on the exsiting stuff.

When you say "document the default value of the options parameter", is it sufficient to document to default values of the relvent options type? Or is there some `/** */` magic to add a default value to a function arg?

UFOMelkor replied:
Normally, we do it like this:
```
   * @param options - Additional file upload options passed as form data
   *                  (default: `{}`)
```

>  but I'm happy to help get the quality level up on the exsiting stuff.

Happy to hear :-)

n3dst4 replied:
Ohhh I see. That's easy enough. I was trying to cram `@default` in there :)

UFOMelkor replied:
To clarify this a bit:

Often there a functions like this

```javascript
/**
 * @param {object} [options] Additional options
 * @param {string} [options.bar=baz] Some description
 */
function foo({bar = 'baz'} = {}) {}
```

and we document it like this.

```typescript
/**
 * @param options - Additional options
 *                  (default: `{}`)
 */
function foo(options?: Partial<Options>): void;

interface Options {
    /**
     * Some description
     * @defaultValue `'baz'`
     */
    bar: string;
}

n3dst4 replied:
So I think we're ready for another review. All the previous items have been addressed, plus I've added more options types to replace those which were previously `Record<string, any>`.

I've also broken up the result types, so `BrowseResult` is different to `ConfigurePathResult`.

Finally, and this may be the controversial bit - I've created a helper type called `PromisedReturnType` which helps when you're forwarding the result of a promise. There are other ways of doing this, including just hard-coding the promised type, but this feels like the most correct solution. If I have over-engineered this just for sake of making `_onDrop` cleaner I will revert those changes :)

UFOMelkor reviewed (changes requested):
That was quite easier to review with the (nearly) correct order :-)

Unfortunately, that means that I can find more to notice ;-)

Thanks for your patience!

n3dst4 commented on src/foundry/foundry.js/applications/filePicker.d.ts:
```diff
@@ -29,27 +39,28 @@ declare class FilePicker<P extends FilePicker.Options = FilePicker.Options> exte
+  extensions: string[] | undefined;
 
   protected _loaded: boolean;
 
```
Moved.

n3dst4 commented on src/foundry/foundry.js/applications/filePicker.d.ts:
```diff
@@ -29,27 +39,28 @@ declare class FilePicker<P extends FilePicker.Options = FilePicker.Options> exte
+  extensions: string[] | undefined;
 
   protected _loaded: boolean;
 
```
done

n3dst4 commented:
Sorry, I forgot to mark the last comment as resolved. 

So I think we're ready for another review. All the previous items have been addressed, plus I've added more options types to replace those which were previously Record<string, any>.

I've also broken up the result types, so BrowseResult is different to ConfigurePathResult.

Finally, and this may be the controversial bit - I've created a helper type called PromisedReturnType which helps when you're forwarding the result of a promise. There are other ways of doing this, including just hard-coding the promised type, but this feels like the most correct solution. If I have over-engineered this just for sake of making _onDrop cleaner I will revert those changes :)

ghost91- commented on src/types/helperTypes.ts:
```diff
@@ -122,3 +122,15 @@ export type DataSourceForPlaceable<P extends PlaceableObject> = P extends Placea
+/**
+ * If `ReturnType<T>` is `Promise<TResult>` then `TResult`; otherwise `ReturnType<T>`.
+ */
+export type PromisedReturnType<T extends (...args: any) => any> = ReturnType<T> extends Promise<infer TResult>
```
Isn't this the same as
```ts
export type PromisedReturnType<T extends (...args: any) => any> = PromisedType<ReturnType<T>>;
```
?

n3dst4 replied:
You're right, I don't know what I was thinking.

n3dst4 commented:
Yes, that extra helper type was pointless. I've move `PromisedType` into `utils.d.ts` and in doing so I no longer need an import in `filePicker.d.ts` so I've reverted the `declare global` I'd added there, so the diff is less catastrophic.

n3dst4 commented:
Oh and thanks to the link to the contrib guidelines, I hadn't seen that.

ghost91- commented on src/foundry/foundry.js/applications/filePicker.d.ts:
```diff
@@ -117,64 +159,75 @@ declare class FilePicker<P extends FilePicker.Options = FilePicker.Options> exte
-  render(force?: boolean, options?: Application.RenderOptions): this | void | Promise<FilePicker.Result>;
+  render(force?: boolean, options?: Application.RenderOptions): ReturnType<this['browse']> | void;
 
   /**
    * Activate listeners to handle user interactivity for the FilePicker UI
```
this doc comment should be replaced by `/** @override */`

ghost91- reviewed (changes requested):
Looks really good, thanks for the work. I only found one more minor thing :)

n3dst4 commented:
Done! Thanks for the review.
